# Exploring eBay Car Sales Data

In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
The original dataset isn't available on Kaggle anymore

In [1]:
import pandas as pd

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

# Clean column names

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos.rename(columns={'yearOfRegistration':'registration_year', 'monthOfRegistration':'registration_month', 'notRepairedDamage':'unrepaired_damage', 'dateCreated':'ad_created', 'dateCrawled':'ad_crawled', 'offerType':'offer_type', 'abtest':'ab_test', 'vehicleType':'vehicle_type', 'gearbox':'gear_box','powerPS':'power_PS', 'fuelType':'fuel_type', 'nrOfPictures':'nr_of_pictures', 'postalCode':'postal_code', 'lastSeen':'last_seen'}, inplace=True)

In [8]:
autos.columns

Index(['ad_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [9]:
autos.head()

ad_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Python prefers snakecase compared to camelcase, which is why we had to rename the following columns.

# Initial exploration and cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

1. Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
2. Examples of numeric data stored as text which can be cleaned and converted.

In [10]:
#Retrieve both categorical and numeric columns
autos.describe(include='all')

ad_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [11]:
autos = autos.drop(['seller','offer_type','nr_of_pictures'], axis=1)
autos.head()

ad_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price  ab_test vehicle_type  registration_year   gear_box  power_PS  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   

    model   odometer  registration_month fuel_type       brand  \
0  andere  150,000km                   3       lpg     peugeot   
1     7er  150,000km                   6    benzin         bmw   
2    golf   70,000km                   7    benzin  volkswagen   
3  fortwo   70,000km                   6    benzin       smart   
4   focus  150,000km                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [12]:
#Rename any non-numeric characters and convert to int dtype
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'].astype(int)

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int32

In [13]:
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'].astype(int)

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 50000, dtype: int32

In [14]:
#Rename odometer to odometer_km
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)

# Exploring the odometer and price columns

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns.

In [15]:
#View min/max/median/mean etc
autos['price'].astype(int).describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
autos['price'].astype(int).sort_index(ascending=True).head(20)

0      5000
1      8500
2      8990
3      4350
4      1350
5      7900
6       300
7      1990
8       250
9       590
10      999
11      350
12     5299
13     1350
14     3999
15    18900
16      350
17     5500
18      300
19     4150
Name: price, dtype: int32

The price varies from a minimum of 300 to a maximum of 18900. We will set a range of 1000 to 10000 to make it more realistic

In [17]:
autos[autos['price'].astype(int).between(1000,10000)]

ad_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49992  2016-03-10 19:37:38              Fiat_Grande_Punto_1.4_T_Jet_16V_Sport   
49993  2016-03-15 18:47:35           Audi_A3__1_8l__Silber;_schoenes_Fahrzeug   
49994  2016-03-22 17:36:42  Audi_A6__S6__Avant_4.2_quattro_eventuell_Tausc...   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

      price  ab_test vehicle_type  registration_year   gear_box  power_PS  \
0      5000  control          bus               2004    manuell       158   
1      8500  control    limousine               1997  automatik       286   
2      8990     test    limousine               2009    manuell       102   
3      4350  control   kleinwagen               2007  automatik        71   
4      1350     test        kombi               2003    manuell         0   
...     ...      ...          ...                ...        ...       ...   
49992  4800  control   kleinwagen               2009    manuell       120   
49993  1650  control   kleinwagen               1997    manuell         0   
49994  5000  control        kombi               2001  automatik       299   
49996  1980  control       cabrio               1996    manuell        75   
49999  1250  control    limousine               1996    manuell       101   

        model odometer_km  registration_month fuel_type       brand  \
0      andere      150000                   3       lpg     peugeot   
1         7er      150000                   6    benzin         bmw   
2        golf       70000                   7    benzin  volkswagen   
3      fortwo       70000                   6    benzin       smart   
4       focus      150000                   7    benzin        ford   
...       ...         ...                 ...       ...         ...   
49992  andere      125000                   9       lpg        fiat   
49993     NaN      150000                   7    benzin        audi   
49994      a6      150000                   1    benzin        audi   
49996   astra      150000                   5    benzin        opel   
49999  vectra      150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49992              nein  2016-03-10 00:00:00        68642  2016-03-13 01:44:51  
49993               NaN  2016-03-15 00:00:00        65203  2016-04-06 19:46:53  
49994              nein  2016-03-22 00:00:00        46537  2016-04-06 08:16:39  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[30567 rows x 17 columns]

Repeat the same steps for the odometer column

In [18]:
autos['odometer_km'].astype(int).describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
autos['odometer_km'].astype(int).sort_index(ascending=True).head(20)

0     150000
1     150000
2      70000
3      70000
4     150000
5     150000
6     150000
7     150000
8     150000
9     150000
10    150000
11    150000
12     50000
13    150000
14    150000
15     80000
16    150000
17    150000
18    150000
19    150000
Name: odometer_km, dtype: int32

No outliers observed for the above.

In [20]:
autos[autos['odometer_km'].astype(int).between(0,150000)]

ad_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price  ab_test vehicle_type  registration_year   gear_box  power_PS  \
0       5000  control          bus               2004    manuell       158   
1       8500  control    limousine               1997  automatik       286   
2       8990     test    limousine               2009    manuell       102   
3       4350  control   kleinwagen               2007  automatik        71   
4       1350     test        kombi               2003    manuell         0   
...      ...      ...          ...                ...        ...       ...   
49995  24900  control    limousine               2011  automatik       239   
49996   1980  control       cabrio               1996    manuell        75   
49997  13200     test       cabrio               2014  automatik        69   
49998  22900  control        kombi               2013    manuell       150   
49999   1250  control    limousine               1996    manuell       101   

        model odometer_km  registration_month fuel_type       brand  \
0      andere      150000                   3       lpg     peugeot   
1         7er      150000                   6    benzin         bmw   
2        golf       70000                   7    benzin  volkswagen   
3      fortwo       70000                   6    benzin       smart   
4       focus      150000                   7    benzin        ford   
...       ...         ...                 ...       ...         ...   
49995      q5      100000                   1    diesel        audi   
49996   astra      150000                   5    benzin        opel   
49997     500        5000                  11    benzin        fiat   
49998      a3       40000                  11    diesel        audi   
49999  vectra      150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

# Exploring the date columns

Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively.

In [21]:
#Select the first 10 characters in each column
autos[['ad_crawled','ad_created','last_seen']][:10]

ad_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
5  2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21
6  2016-03-20 17:55:21  2016-03-20 00:00:00  2016-03-23 02:48:59
7  2016-03-16 18:55:19  2016-03-16 00:00:00  2016-04-07 03:17:32
8  2016-03-22 16:51:34  2016-03-22 00:00:00  2016-03-26 18:18:10
9  2016-03-16 13:47:02  2016-03-16 00:00:00  2016-04-06 10:46:35

In [22]:
#Include missing values in the distribution and view as percentages
autos['ad_crawled'].value_counts(normalize=True, dropna=False).sort_index()

ad_crawled
2016-03-05 14:06:30    0.00002
2016-03-05 14:06:40    0.00002
2016-03-05 14:07:04    0.00002
2016-03-05 14:07:08    0.00002
2016-03-05 14:07:21    0.00002
                        ...   
2016-04-07 14:30:09    0.00002
2016-04-07 14:30:26    0.00002
2016-04-07 14:36:44    0.00002
2016-04-07 14:36:55    0.00002
2016-04-07 14:36:56    0.00002
Name: proportion, Length: 48213, dtype: float64

All dates above were crawled at the same rate at 2%.

In [23]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

ad_created
2015-06-11 00:00:00    0.00002
2015-08-10 00:00:00    0.00002
2015-09-09 00:00:00    0.00002
2015-11-10 00:00:00    0.00002
2015-12-05 00:00:00    0.00002
                        ...   
2016-04-03 00:00:00    0.03892
2016-04-04 00:00:00    0.03688
2016-04-05 00:00:00    0.01184
2016-04-06 00:00:00    0.00326
2016-04-07 00:00:00    0.00128
Name: proportion, Length: 76, dtype: float64

Highest percentage is 11.84% on 3rd April 2016 at midnight.

In [24]:
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index()

last_seen
2016-03-05 14:45:46    0.00002
2016-03-05 14:46:02    0.00002
2016-03-05 14:49:34    0.00002
2016-03-05 15:16:11    0.00002
2016-03-05 15:16:47    0.00002
                        ...   
2016-04-07 14:58:44    0.00006
2016-04-07 14:58:45    0.00002
2016-04-07 14:58:46    0.00002
2016-04-07 14:58:48    0.00006
2016-04-07 14:58:50    0.00008
Name: proportion, Length: 39481, dtype: float64

In [25]:
#View the distribution for the registration year
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The minimum value is 1000, before cars were invented.
The maximum value is 9999, many years into the future.

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

# Dealing with Incorrect Registration Year Data

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [26]:
#Set the year range from 1900 to 2016
autos = autos[autos['registration_year'].between(1900,2016)]

In [27]:
#Decide which the highest and lowest acceptable values are
autos['registration_year'].value_counts(normalize=True)

registration_year
2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
          ...   
1929    0.000021
1952    0.000021
1939    0.000021
1938    0.000021
1953    0.000021
Name: proportion, Length: 78, dtype: float64

Most vehicles were registered in the year 2000.

In [28]:
autos['registration_year'].describe()

count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64

# Exploring price by brand

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the brand column.

In [29]:
#Select the first 10 brands
top_brands_counts = autos['brand'].value_counts(normalize=True).head(10)
top_brands = top_brands_counts.index
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object', name='brand')


In [30]:
#Calculate the mean price of each brand assigned
top_brands_mean_price = {}

for b in top_brands:
    selected_rows = autos[autos['brand'] == b]
    mean_price = selected_rows['price'].astype(int).mean()
    top_brands_mean_price[b] = mean_price
    
print(top_brands_mean_price)

{'volkswagen': 6516.457597173145, 'bmw': 8334.645155185466, 'opel': 5252.61655437921, 'mercedes_benz': 30317.447816593885, 'audi': 9093.65003615329, 'ford': 7263.015811455847, 'renault': 2395.4164467897976, 'peugeot': 3039.4682651622, 'fiat': 2711.8011272141707, 'seat': 4296.492554410081}


# Storing aggregate data in a dataframe

We can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To do this, we'll need to learn two pandas methods:

1. pandas series constructor
2. pandas dataframe constructor

In [31]:
#Calculate the mean mileage
top_brands_mean_mileage = {}

for b in top_brands:
    selected_rows = autos[autos["brand"] == b]
    mean_mileage = selected_rows["odometer_km"].astype(int).mean()
    top_brands_mean_mileage[b] = mean_mileage
    
print(top_brands_mean_mileage)

{'volkswagen': 128730.36906164115, 'bmw': 132434.70855412565, 'opel': 129227.14148219442, 'mercedes_benz': 130860.26200873363, 'audi': 129287.78018799711, 'ford': 124046.83770883054, 'renault': 128183.81706244503, 'peugeot': 127136.81241184767, 'fiat': 116553.94524959743, 'seat': 121563.57388316152}


In [32]:
#Convert both dictionaries to series constructors
mean_mileage_series = (pd.Series(top_brands_mean_mileage)
                       .sort_values(ascending=True))

print(mean_mileage_series)

fiat             116553.945250
seat             121563.573883
ford             124046.837709
peugeot          127136.812412
renault          128183.817062
volkswagen       128730.369062
opel             129227.141482
audi             129287.780188
mercedes_benz    130860.262009
bmw              132434.708554
dtype: float64


Fiat displays the lowest mileage whereas BMW is the highest.

In [33]:
mean_price_series = pd.Series(top_brands_mean_price)
print(mean_price_series)

volkswagen        6516.457597
bmw               8334.645155
opel              5252.616554
mercedes_benz    30317.447817
audi              9093.650036
ford              7263.015811
renault           2395.416447
peugeot           3039.468265
fiat              2711.801127
seat              4296.492554
dtype: float64


Mercedes Benz is the most expensive while Renault is the less expensive

In [34]:
#Convert both dictionaries to dataframe constructors
brand_dataframe = pd.DataFrame(mean_mileage_series, columns=["mean_mileage"])
brand_dataframe

mean_mileage
fiat           116553.945250
seat           121563.573883
ford           124046.837709
peugeot        127136.812412
renault        128183.817062
volkswagen     128730.369062
opel           129227.141482
audi           129287.780188
mercedes_benz  130860.262009
bmw            132434.708554

In [35]:
brand_dataframe["mean_price"] = mean_price_series
brand_dataframe

mean_mileage    mean_price
fiat           116553.945250   2711.801127
seat           121563.573883   4296.492554
ford           124046.837709   7263.015811
peugeot        127136.812412   3039.468265
renault        128183.817062   2395.416447
volkswagen     128730.369062   6516.457597
opel           129227.141482   5252.616554
audi           129287.780188   9093.650036
mercedes_benz  130860.262009  30317.447817
bmw            132434.708554   8334.645155

# Conclusion

Based on our analysis, we can conclude that Renault while is the cheapest has comparable mileage compared to other brands and thus will be a good buy for a first time purchaser.